cstar/base/local_file_stats.py

```diff
@@ -0,0 +1,229 @@
+
+    Parameters
+    ----------
+    paths : list of Path
+        List of file paths to track. All paths must be in the same parent directory.
+    stats : dict of Path to os.stat_result, optional
+        Optional precomputed stat results. If not provided, they will be computed lazily.
+    hashes : dict of Path to str, optional
+        Optional precomputed SHA-256 hashes. If not provided, they will be computed lazily.
+
+    Attributes
+    ----------
+    paths : list of Path
+        The list of tracked file paths.
+    parent_dir : Path
+        The common parent directory of all paths.
+    stats : dict of Path to os.stat_result
+        Mapping of paths to their `os.stat()` results. Evaluated lazily.
+    hashes : dict of Path to str
+        Mapping of paths to their SHA-256 hash strings. Evaluated lazily.
+
+    Raises
+    ------
+    ValueError
+        If the provided paths are not all in the same directory
+        or if the keys of the `stats` or `hashes` dictionaries do not exactly
+        match the resolved set of `paths`.
+
+    Examples
+    --------
+    >>> files = [Path("data/a.txt"), Path("data/b.txt")]
+    >>> tracker = LocalFileStatistics(files)
+    >>> tracker.stats[files[0]]
+    os.stat_result(...)
+    >>> tracker.hashes[files[0]]
+    'e3b0c44298fc1c149afbf4c8996fb924...'
+    """
+
+    def __init__(
+        self,
+        paths: list[Path],
+        stats: Optional[dict[Path, os.stat_result]] = None,
+        hashes: Optional[dict[Path, str]] = None,
+    ):
+        """Initialize the LocalFileStatistics object.
+
+        Stores a list of local file paths and optionally precomputed metadata
+        such as file stat results and SHA-256 hashes. All paths must be located
+        within the same parent directory. Metadata is lazily computed if not
+        provided.
+
+        Parameters
+        ----------
+        paths : list of Path
+            List of file paths to track. All paths must reside in the same directory.
+        stats : dict of Path to os.stat_result, optional
+            Optional precomputed file stat results. If provided, must include all
+            tracked files and match `paths` exactly after resolution.
+        hashes : dict of Path to str, optional
+            Optional precomputed SHA-256 hashes. If provided, must include all
+            tracked files and match `paths` exactly after resolution.
+
+        Raises
+        ------
+        ValueError
+            If the provided paths are not all in the same directory or if the keys
+            of `stats` or `hashes` do not match the resolved paths.
+        """
+
+        self.paths = [p.absolute() for p in paths]
+
+        if not all([d.parent == self.paths[0].parent for d in self.paths]):
+            raise ValueError(
+                "LocalFileStatistics requires paths to exist in a common directory"
+            )
+
+        self.parent_dir = paths[0].parent
+
+        if stats is not None:
```

Comment:
Same deal with nulls. Look for a way to process additions immediately to avoid having a null stats and hashes member (same as comment about instantiating the LocalFileStats to an empty version).

If we're exposing a list, like "self.files = []` and allowing people to modify it at-will, we put ourselves in this position where we can't trust our member variables. If we hide that as `self._files = []` we can expose 

```python
def add_files(self, flist: list[Path]) -> None:
   if not validate(f):
      raise ValueError
   self._files.extend(f.absolute() for f in flist)
```

When I notice that I'm making myself do repeated checks for nulls, i always start to ask "how can I avoid doing this everywhere?"